File: phases/13-concurrent-and-parallel/07-atomics-cas-aba-problem/code/main.cpp
// Phase 13, Lesson 07 — Atomics, CAS, ABA Problem (C++)
// Demonstrates: atomic counters, lock-free Treiber stack, the ABA problem
// with a node-recycling allocator, and tagged-pointer ABA prevention.
//
// Compile:  clang++ -std=c++17 -O2 -pthread main.cpp -o atomic_lesson_cpp
// Run:      ./atomic_lesson_cpp

#include <atomic>
#include <chrono>
#include <cassert>
#include <iostream>
#include <thread>
#include <vector>

// ============================================================================
// Utility: stopwatch
// ============================================================================

struct Stopwatch {
    using Clock = std::chrono::high_resolution_clock;
    Clock::time_point start_ = Clock::now();

    double elapsed_ms() const {
        auto end = Clock::now();
        return std::chrono::duration<double, std::milli>(end - start_).count();
    }
};

// ============================================================================
// Step 1: Atomic Counter
// ============================================================================

class AtomicCounter {
    std::atomic<unsigned long long> val_{0};
public:
    void increment() { val_.fetch_add(1, std::memory_order_relaxed); }
    unsigned long long get() const { return val_.load(std::memory_order_relaxed); }
};

class MutexCounter {
    mutable std::mutex mtx_;
    unsigned long long val_{0};
public:
    void increment() {
        std::lock_guard<std::mutex> lock(mtx_);
        ++val_;
    }
    unsigned long long get() const {
        std::lock_guard<std::mutex> lock(mtx_);
        return val_;
    }
};

void bench_counter() {
    constexpr int N = 1'000'000;
    constexpr int THREADS = 8;

    // Atomic counter
    AtomicCounter ac;
    Stopwatch sw;
    std::vector<std::thread> threads;
    for (int i = 0; i < THREADS; ++i)
        threads.emplace_back([&]() { for (int j = 0; j < N; ++j) ac.increment(); });
    for (auto& t : threads) t.join();
    double atomic_ms = sw.elapsed_ms();
    std::cout << "  Atomic counter: " << (N * THREADS) << " in " << atomic_ms
              << " ms (final=" << ac.get() << ")\n";

    // Mutex counter
    MutexCounter mc;
    threads.clear();
    sw = Stopwatch{};
    for (int i = 0; i < THREADS; ++i)
        threads.emplace_back([&]() { for (int j = 0; j < N; ++j) mc.increment(); });
    for (auto& t : threads) t.join();
    double mutex_ms = sw.elapsed_ms();
    std::cout << "  Mutex counter:  " << (N * THREADS) << " in " << mutex_ms
              << " ms (final=" << mc.get() << ")\n";

    std::cout << "  Speedup: " << (mutex_ms / atomic_ms) << "x\n";
}

// ============================================================================
// Step 2: Lock-Free Stack (Treiber Stack) — raw pointers, no ABA protection
// ============================================================================

struct Node {
    int value;
    Node* next;
};

class LockFreeStack {
    std::atomic<Node*> head_{nullptr};
public:
    void push(Node* node) {
        Node* old = head_.load(std::memory_order_acquire);
        do {
            node->next = old;
        } while (!head_.compare_exchange_weak(old, node,
                                               std::memory_order_release,
                                               std::memory_order_relaxed));
    }

    Node* pop() {
        Node* old = head_.load(std::memory_order_acquire);
        while (old) {
            Node* nxt = old->next;
            if (head_.compare_exchange_weak(old, nxt,
                                            std::memory_order_release,
                                            std::memory_order_relaxed)) {
                return old;
            }
        }
        return nullptr;
    }

    // Expose head_ for the ABA demonstration below.
    // In production code this would be private; we expose it here to
    // simulate the thread interleaving that causes the ABA bug.
    std::atomic<Node*>& debug_head() { return head_; }
};

void test_lockfree_stack() {
    LockFreeStack stack;
    constexpr int N = 1000;

    std::vector<std::thread> pushers;
    for (int start = 0; start < N; start += 250) {
        pushers.emplace_back([&, start]() {
            for (int i = start; i < start + 250; ++i) {
                stack.push(new Node{i, nullptr});
            }
        });
    }
    for (auto& t : pushers) t.join();

    std::vector<int> popped;
    while (Node* n = stack.pop()) {
        popped.push_back(n->value);
        delete n;
    }

    std::sort(popped.begin(), popped.end());
    assert(popped.size() == N);
    for (int i = 0; i < N; ++i) assert(popped[i] == i);
    std::cout << "  Lock-free stack: " << popped.size() << " values OK\n";
}

// ============================================================================
// Step 3: The ABA Problem
// ============================================================================
// We use a single-slot recycling allocator so that address reuse is
// deterministic — the ABA bug always manifests.
//
// Sequence:
//   1. Thread T1: reads head = node A, computes new_head = A->next = B.
//   2. Thread T2: pops A, pops B, recycles A's address.
//   3. Thread T2: allocates C at A's recycled address, pushes C.
//   4. Thread T1: CAS(&head, A, B) — SUCCEEDS because head == C == A's address!
//   5. head now points to B — but B was freed! Data corruption.

Node* recycled = nullptr;

void recycle_node(Node* n) {
    recycled = n;
}

Node* alloc_node(int value) {
    if (recycled) {
        Node* p = recycled;
        recycled = nullptr;
        p->value = value;
        p->next = nullptr;
        return p;
    }
    return new Node{value, nullptr};
}

void demonstrate_aba() {
    std::cout << "\n=== ABA Problem Demonstration ===\n\n";

    // Step 0: allocate A and B, build stack: A -> B -> null
    Node* A = alloc_node(1);
    Node* B = alloc_node(2);
    A->next = B;
    B->next = nullptr;

    std::atomic<Node*> head{A};

    std::cout << "Initial stack: head -> A(" << A << ", val=1)"
              << " -> B(" << B << ", val=2) -> null\n\n";

    // Step 1: Thread T1 reads head = A, plans CAS(&head, A, A->next=B)
    Node* observed = head.load(std::memory_order_acquire);
    Node* new_head = observed->next;  // B
    std::cout << "[T1] load(head) = " << observed
              << "  (will CAS(&head, " << observed << ", " << new_head << "))\n";

    // Step 2: Thread T2 pops A, pops B, recycles A
    head.store(B, std::memory_order_release);          // pop A
    std::cout << "[T2] pop(A): head -> B\n";

    head.store(nullptr, std::memory_order_release);    // pop B
    std::cout << "[T2] pop(B): head -> null\n";

    recycle_node(A);   // A's address is now available for reuse
    std::cout << "[T2] recycle(A) — address " << A << " is freed\n";

    // Step 3: Thread T2 allocates C at A's recycled address, pushes C
    Node* C = alloc_node(3);  // same address as A!
    head.store(C, std::memory_order_release);
    std::cout << "[T2] alloc(C) at " << C << " (reuses A's address, val=3)\n";
    std::cout << "[T2] push(C): head -> C\n\n";

    // Step 4: Thread T1 resumes — CAS(&head, observed=A, new_head=B)
    // head is C, which has the same address as A.
    // (observed) == (C) == address of A.
    bool cas_result = head.compare_exchange_strong(observed, new_head);
    std::cout << "[T1] CAS(&head, " << observed << ", " << new_head << ")"
              << " = " << std::boolalpha << cas_result << "\n";

    if (cas_result) {
        Node* current_head = head.load();
        std::cout << "[T1] head now = " << current_head
                  << " which is FREED MEMORY (B was deleted)!\n";
        std::cout << "[T1] ABA BUG CONFIRMED — data structure corrupted.\n";
    } else {
        std::cout << "[T1] CAS failed — no bug this run.\n";
    }

    // Cleanup (in a real program we would leak C here — for the demo, delete it).
    delete C;
    delete B;

    std::cout << "\n=== End of ABA Demo ===\n";
}

// ============================================================================
// Step 4: Tagged Pointer (ABA Solution)
// ============================================================================
// Embed a version counter in the lowest 3 bits of the pointer.
// Pointers are at least 8-byte aligned → bottom 3 bits are always 0.
// Every CAS increments the tag. Even if the address cycles, the tag
// won't match → CAS fails → safe retry.

constexpr uintptr_t TAG_MASK = 0x7;  // 3 bits
constexpr uintptr_t PTR_MASK = ~TAG_MASK;

inline uintptr_t pack(Node* ptr, uintptr_t tag) {
    return (reinterpret_cast<uintptr_t>(ptr) & PTR_MASK) | (tag & TAG_MASK);
}

inline std::pair<Node*, uintptr_t> unpack(uintptr_t val) {
    return {reinterpret_cast<Node*>(val & PTR_MASK), val & TAG_MASK};
}

class TaggedStack {
    std::atomic<uintptr_t> head_{0};  // packed: address | tag
public:
    void push(Node* node) {
        uintptr_t old = head_.load(std::memory_order_acquire);
        do {
            auto [ptr, tag] = unpack(old);
            node->next = ptr;
            uintptr_t new_tag = (tag + 1) & TAG_MASK;
            uintptr_t desired = pack(node, new_tag);
            if (head_.compare_exchange_weak(old, desired,
                                            std::memory_order_release,
                                            std::memory_order_relaxed))
                return;
        } while (true);
    }

    Node* pop() {
        uintptr_t old = head_.load(std::memory_order_acquire);
        while (true) {
            auto [ptr, tag] = unpack(old);
            if (!ptr) return nullptr;
            Node* next = ptr->next;
            uintptr_t new_tag = (tag + 1) & TAG_MASK;
            uintptr_t desired = pack(next, new_tag);
            if (head_.compare_exchange_weak(old, desired,
                                            std::memory_order_release,
                                            std::memory_order_relaxed))
                return ptr;
        }
    }

    std::atomic<uintptr_t>& debug_head() { return head_; }
    uintptr_t debug_head_val() const { return head_.load(); }
};

void test_tagged_stack() {
    // Reproduce the ABA scenario but with tagged pointer protection.
    // The CAS should FAIL and retry, preventing corruption.

    std::cout << "\n=== Tagged Pointer — ABA Prevention ===\n\n";

    Node* A = alloc_node(1);
    Node* B = alloc_node(2);
    A->next = B;

    TaggedStack stack;
    stack.push(B);
    stack.push(A);
    // Stack: A -> B -> null, head = pack(A, tag=0)

    uintptr_t head_val_before = stack.debug_head_val();
    auto [ptr_before, tag_before] = unpack(head_val_before);
    std::cout << "Initial: head = pack(" << ptr_before << ", tag=" << tag_before << ")\n";

    // Simulate the interleaving
    // Step 1: read head
    uintptr_t observed = head_val_before;
    auto [observed_ptr, observed_tag] = unpack(observed);
    Node* observed_next = observed_ptr->next;  // B
    std::cout << "[T1] load(head) = pack(" << observed_ptr
              << ", tag=" << observed_tag << ")\n";

    // Step 2: pop A, pop B, recycle A
    stack.pop();                    // pops A
    stack.pop();                    // pops B (stack empty now)

    // head is now pack(nullptr, tag=2) — tag incremented twice
    uintptr_t after_pop = stack.debug_head_val();
    std::cout << "[T2] after 2 pops: head = " << after_pop << "\n";

    // Recycled A will be reused
    Node* C = alloc_node(3);        // same address as A
    recycle_node(nullptr);          // reset recycler

    // Try to push C with the tagged stack (which increments tag)
    stack.push(C);
    uintptr_t after_push = stack.debug_head_val();
    auto [push_ptr, push_tag] = unpack(after_push);
    std::cout << "[T2] push(C at " << push_ptr << "): head = pack("
              << push_ptr << ", tag=" << push_tag << ")\n";

    // Step 3: T1 tries CAS with the old tag
    // In a real tagged stack, T1 would use the full head value (including tag)
    // But here we manually test: CAS(&head_, observed, pack(B, tag+1))
    // observed has tag=0, but head now has tag=3
    // They DON'T match → CAS fails → no corruption
    uintptr_t new_val = pack(observed_next, (observed_tag + 1) & TAG_MASK);
    bool cas_result = stack.debug_head().compare_exchange_strong(observed, new_val);

    std::cout << "[T1] CAS(&head, " << observed << ", " << new_val << ")"
              << " = " << std::boolalpha << cas_result << "\n";

    if (!cas_result) {
        auto [final_ptr, final_tag] = unpack(observed);
        std::cout << "[T1] CAS correctly FAILED — observed updated to pack("
                  << final_ptr << ", tag=" << final_tag << ")\n";
        std::cout << "Tagged pointer prevented ABA! ✓\n";
    } else {
        std::cout << "[T1] CAS succeeded (unexpected — tag collision)\n";
    }

    // Drain remaining nodes
    delete B;
    while (stack.pop()) {}
}

// ============================================================================
// Main
// ============================================================================

int main() {
    std::cout << "=== Phase 13.07: Atomics, CAS, ABA Problem (C++) ===\n\n";

    std::cout << "--- Step 1: Atomic Counter (FAA vs Mutex) ---\n";
    std::cout << "  8 threads x 1,000,000 increments each\n";
    bench_counter();

    std::cout << "\n--- Step 2: Lock-Free Stack ---\n";
    test_lockfree_stack();

    std::cout << "\n--- Step 3: The ABA Problem ---\n";
    demonstrate_aba();

    std::cout << "\n--- Step 4: Tagged Pointer Solution ---\n";
    test_tagged_stack();

    std::cout << "\n=== All steps completed. ===\n";
    return 0;
}

File: phases/13-concurrent-and-parallel/07-atomics-cas-aba-problem/code/main.rs
// Phase 13, Lesson 07 — Atomics, CAS, ABA Problem
// Demonstrates: atomic counters, lock-free Treiber stack via CAS,
// the ABA problem (conceptual in Rust), and tagged-pointer ABA prevention.
//
// Compile:  rustc main.rs -o atomic_lesson
// Run:      ./atomic_lesson

use std::marker::PhantomData;
use std::sync::atomic::{AtomicPtr, AtomicUsize, Ordering};
use std::sync::{Arc, Mutex};
use std::thread;
use std::time::Instant;

// ============================================================================
// Step 1: Atomic Counter
// ============================================================================
// Compare AtomicUsize::fetch_add (FAA) against Mutex-protected increment.

struct AtomicCounter(AtomicUsize);

impl AtomicCounter {
    fn new() -> Self {
        Self(AtomicUsize::new(0))
    }

    fn increment(&self) -> usize {
        self.0.fetch_add(1, Ordering::Relaxed)
    }

    fn get(&self) -> usize {
        self.0.load(Ordering::Relaxed)
    }
}

struct MutexCounter {
    inner: Mutex<usize>,
}

impl MutexCounter {
    fn new() -> Self {
        Self { inner: Mutex::new(0) }
    }

    fn increment(&self) {
        *self.inner.lock().unwrap() += 1;
    }

    fn get(&self) -> usize {
        *self.inner.lock().unwrap()
    }
}

fn bench_counter() {
    const N: usize = 1_000_000;
    const THREADS: usize = 8;

    let atomic = AtomicCounter::new();
    let start = Instant::now();
    thread::scope(|s| {
        for _ in 0..THREADS {
            s.spawn(|| {
                for _ in 0..N {
                    atomic.increment();
                }
            });
        }
    });
    let atomic_dur = start.elapsed();
    println!("  Atomic counter: {} in {:?} (final={})",
        N * THREADS, atomic_dur, atomic.get());

    let mutex = MutexCounter::new();
    let start = Instant::now();
    thread::scope(|s| {
        for _ in 0..THREADS {
            s.spawn(|| {
                for _ in 0..N {
                    mutex.increment();
                }
            });
        }
    });
    let mutex_dur = start.elapsed();
    println!("  Mutex counter:  {} in {:?} (final={})",
        N * THREADS, mutex_dur, mutex.get());

    println!("  Speedup: {:.2}x", mutex_dur.as_secs_f64() / atomic_dur.as_secs_f64());
    println!("  Note: Ordering::Relaxed is safe — no cross-thread ordering needed.");
}

// ============================================================================
// Step 2: CAS-based Lock-Free Stack (Treiber Stack)
// ============================================================================

struct Node<T> {
    value: T,
    next: *mut Node<T>,
}

unsafe impl<T: Send> Send for Node<T> {}
unsafe impl<T: Send> Sync for Node<T> {}

struct LockFreeStack<T> {
    head: AtomicPtr<Node<T>>,
}

unsafe impl<T: Send> Send for LockFreeStack<T> {}
unsafe impl<T: Send> Sync for LockFreeStack<T> {}

impl<T> Drop for LockFreeStack<T> {
    fn drop(&mut self) {
        let mut ptr = self.head.load(Ordering::Relaxed);
        while !ptr.is_null() {
            let node = unsafe { Box::from_raw(ptr) };
            ptr = node.next;
        }
    }
}

impl<T> LockFreeStack<T> {
    fn new() -> Self {
        Self { head: AtomicPtr::new(std::ptr::null_mut()) }
    }

    fn push(&self, value: T) {
        let node = Box::into_raw(Box::new(Node {
            value,
            next: std::ptr::null_mut(),
        }));
        loop {
            let head = self.head.load(Ordering::Acquire);
            unsafe { (*node).next = head; }
            if self.head
                .compare_exchange(head, node, Ordering::Release, Ordering::Relaxed)
                .is_ok()
            {
                break;
            }
        }
    }

    fn pop(&self) -> Option<T> {
        loop {
            let head = self.head.load(Ordering::Acquire);
            if head.is_null() {
                return None;
            }
            let next = unsafe { (*head).next };
            if self.head
                .compare_exchange(head, next, Ordering::Release, Ordering::Relaxed)
                .is_ok()
            {
                let node = unsafe { Box::from_raw(head) };
                return Some(node.value);
            }
        }
    }
}

fn test_lockfree_stack() {
    let stack = Arc::new(LockFreeStack::new());
    let values: Vec<i32> = (0..2000).collect();

    let mut handles = vec![];
    for chunk in values.chunks(500) {
        let local = chunk.to_vec();
        let stack_clone = stack.clone();
        handles.push(thread::spawn(move || {
            for &v in &local {
                stack_clone.push(v);
            }
        }));
    }
    for h in handles {
        h.join().unwrap();
    }

    let mut popped = Vec::new();
    while let Some(v) = stack.pop() {
        popped.push(v);
    }

    popped.sort();
    assert_eq!(popped, values);
    println!("  Lock-free stack: {} values pushed + popped correctly", popped.len());
}

// ============================================================================
// Step 3: The ABA Problem (Conceptual in Rust)
// ============================================================================
//
// The ABA problem: CAS checks *address equality* on a pointer. If the address
// cycles A → B → A, CAS cannot distinguish "nothing happened" from
// "something happened and was undone."
//
// In C++ with manual memory management:
//   T1: head = load(A)           → plans CAS(&head, A, A->next=B)
//   T1: (preempted)
//   T2: pop(A), free(A)          → head = B
//   T2: pop(B), free(B)          → head = null
//   T2: alloc(C) at A's addr     → head = C (C at same address as A!)
//   T1: CAS(&head, A, B)         → SUCCEEDS (head == A's address)
//                                → head = B, BUT B WAS FREED!
//
// Rust's borrow checker prevents this exact scenario: you cannot call
// Box::from_raw (free) while any raw pointer to the memory still exists,
// at least not without extensive unsafe that recreates C++ semantics.
//
// The C++ file (main.cpp) in this lesson provides a deterministic,
// runnable demonstration of the ABA bug using a node-recycling allocator.
// Run it to see the corruption in action.

fn aba_conceptual_demo() {
    println!("  Rust's borrow checker prevents the simplest ABA scenario.");
    println!("  See main.cpp for a deterministic ABA demonstration with");
    println!("  a node-recycling allocator that reuses addresses.");
    println!();
    println!("  Key insight: CAS only checks bit-pattern equality.");
    println!("  If pointer address matches but memory has been recycled,");
    println!("  the data structure is corrupted.");
}

// ============================================================================
// Step 4: ABA Solution — Tagged Pointer
// ============================================================================
// Embed a version counter in the lowest 3 bits of the AtomicUsize.
// User-space pointers are at least 8-byte aligned (bottom 3 bits are 0),
// so we can safely co-locate a 3-bit tag.
//
// Each successful CAS increments the tag. Even if the pointer value cycles
// back to an old address, the tag differs → CAS fails → retry.

const TAG_BITS: usize = 3;
const TAG_MASK: usize = (1 << TAG_BITS) - 1;
const PTR_MASK: usize = !TAG_MASK;

fn pack<T>(ptr: *mut T, tag: usize) -> usize {
    (ptr as usize & PTR_MASK) | (tag & TAG_MASK)
}

fn unpack<T>(val: usize) -> (*mut T, usize) {
    let ptr = (val & PTR_MASK) as *mut T;
    let tag = val & TAG_MASK;
    (ptr, tag)
}

struct TaggedStack<T> {
    head: AtomicUsize,
    _phantom: PhantomData<T>,
}

unsafe impl<T: Send> Send for TaggedStack<T> {}
unsafe impl<T: Send> Sync for TaggedStack<T> {}

impl<T> Drop for TaggedStack<T> {
    fn drop(&mut self) {
        let (mut ptr, _): (*mut Node<T>, _) = unpack(self.head.load(Ordering::Relaxed));
        while !ptr.is_null() {
            let node = unsafe { Box::from_raw(ptr) };
            ptr = node.next;
        }
    }
}

impl<T> TaggedStack<T> {
    fn new() -> Self {
        Self {
            head: AtomicUsize::new(pack::<Node<T>>(std::ptr::null_mut(), 0)),
            _phantom: PhantomData,
        }
    }

    fn push(&self, value: T) {
        let node = Box::into_raw(Box::new(Node {
            value,
            next: std::ptr::null_mut(),
        }));
        loop {
            let head_val = self.head.load(Ordering::Acquire);
            let (head_ptr, _tag): (*mut Node<T>, _) = unpack(head_val);
            unsafe { (*node).next = head_ptr; }
            let new_val = pack(node, (_tag + 1) & TAG_MASK);
            if self.head
                .compare_exchange(head_val, new_val, Ordering::Release, Ordering::Relaxed)
                .is_ok()
            {
                break;
            }
        }
    }

    fn pop(&self) -> Option<T> {
        loop {
            let head_val = self.head.load(Ordering::Acquire);
            let (head_ptr, tag): (*mut Node<T>, _) = unpack(head_val);
            if head_ptr.is_null() {
                return None;
            }
            let next_ptr = unsafe { (*head_ptr).next };
            let new_val = pack(next_ptr, (tag + 1) & TAG_MASK);
            if self.head
                .compare_exchange(head_val, new_val, Ordering::Release, Ordering::Relaxed)
                .is_ok()
            {
                let node = unsafe { Box::from_raw(head_ptr) };
                return Some(node.value);
            }
        }
    }
}

fn test_tagged_stack() {
    let stack = TaggedStack::new();
    let n = 10_000;

    thread::scope(|s| {
        s.spawn(|| {
            for i in 0..n {
                stack.push(i);
            }
        });
        s.spawn(|| {
            for i in 0..n {
                stack.push(i + n);
            }
        });
    });

    let mut count = 0;
    while stack.pop().is_some() {
        count += 1;
    }
    assert_eq!(count, 2 * n);
    println!("  Tagged-pointer stack: {} values pushed + popped correctly", count);
    println!("  (3-bit tag prevents ABA — each CAS increments version)");
}

// ============================================================================
// Main — run all four steps sequentially
// ============================================================================

fn main() {
    println!("=== Phase 13.07: Atomics, CAS, ABA Problem ===\n");

    println!("--- Step 1: Atomic Counter (FAA vs Mutex) ---");
    println!("  8 threads x 1,000,000 increments each");
    bench_counter();

    println!();
    println!("--- Step 2: Lock-Free Stack (Treiber Stack) ---");
    println!("  CAS-based push/pop on AtomicPtr<Node<T>>");
    test_lockfree_stack();

    println!();
    println!("--- Step 3: The ABA Problem ---");
    aba_conceptual_demo();

    println!();
    println!("--- Step 4: ABA Solution — Tagged Pointer ---");
    println!("  Version counter in bottom 3 AtomicUsize bits");
    test_tagged_stack();

    println!();
    println!("=== All steps completed. ===");
}
